=Other functions=

==Genome region operations==

Some examples of operations on genome regions files. Although some of the operations can also be achieved by advanced queries, we show here how to use regions files.

Select regions sequenced by CG (with two SNV callers)
{{{
cg select -q 'count($sequenced-cg*, == 1) == 4' \
	NA19240_chrom22.tsv.gz > reg_2.tsv
}}}

Then calculate how large the selected regions are
{{{
cg covered reg2.tsv > reg2.covered
}}}

Select regions representing all missense SNVs
{{{
cg select -q '$impact =="MISSENSE" && $type == "snp" ' \
	NA19240_chrom22.tsv.gz > reg_missense.tsv
}}}

Join both region files
{{{
cg regjoin reg_2.tsv reg_missense.tsv > joined.tsv
}}}

Subtract missense SNVs from the first file
{{{
cg regsubtract reg_2.tsv reg_missense.tsv > subtracted.tsv
}}}

==Primer design==


First select regions for design for all missense SNVs
{{{
cg makeregions reg_missense.tsv 200 > regval_missense.tsv
}}}

Then make primers for them using the experiment name ValidationName and the location of the reference genome
{{{
cg makeprimers regval_missense.tsv ValidationName 600 500 dbdir > primers-valreg.tsv
}}}
