= Annotate =

== Format ==
cg annotate ?options? variantfile resultfile dbfile ...

== Summary ==
Annotate a variant file with region, gene or variant data

== Description ==
Adds new columns with annotation to variantfile.
Each dbfile will add 1 or more columns to the resultfile.
different types of dbfiles are treated differently. The type is determined 
based on the first part of the filename (before the first underscore).
Each column will start with a base name (the part of the filename
after the last underscore)

== Arguments ==
; **variantfile**: file in sft format with variant data
; **resultfile**: resulting file in sft format with new columns added
; **dbfile**: files (sft format) with features used for annotation

== Options ==
; **-near dist**: also annotate variants with the nearest feature in 
the dbfile if it is closer than **dist** to it. A column name_dist will
be added that contains the distance.

== database types ==
; **reg**: regions file that must at least contain the columns chromosome,start,end. 
Variants are checked for overlap with regions in the file.
; **var**: variations file that must at least contain the columns chromosome,start,end,type,ref,alt 
to annotate variants that match the given values. Typically, columns freq and id are present for annotation.
Thus, only variants that match the alleles given in alt will be annotated. If there are multiple alt 
values (separated by commas), the freq annotation will also contain multiple fields (separated by commas).
; **gene**: gene files, currently only overlap is annotated
; **annovar**: use annovar to annotate genes

If a database filename does not start with one of these types, 
it will be considered a regions database.

== database parameters ==
If a file dbfile.opt exists, it will be scanned for database parameters.
It should be a tab separated list, where each line contains a key and a 
value (separated by a tab)

Possible keys are:
; **name**: this will be the base for names of added columns (in stead of extracting it from the filename)
; **fields**: These fields will be extracted from the database and added to the annotated file in stead 
  of the defaults (one or more of name, name2, freq and score, depending on the type and name of the database)

== Category ==
Annotation
