= cg2tsv =

== Format ==
cg cg2tsv ?options? cgvarfile ?cggenefile? outputfile

== Summary ==
Converts data in Complete Genomics format to tsv (tab separated, simple feature table) format.
The command will also sort the tsv appropriately.
The cggenefile is optional. It contains the CG gene annotations, which can be (best) left out as
these annotations (and more) can be done in genomecomb.

== Arguments ==
; **cgvarfile**: Complete genomics var file to be converted. File may be compressed.
; **cggenefile**: Complete genomics gene file to be converted.
; **outfile**: write results to outfile

== Options ==
; **-split**: If 1 (default),  will be on a separate lines, treated mostly as a separate variant
; **-sorted 0/1**: Use 1 if the CG var and gene file are already sorted (so they are not sorted gain first)
; **-ref refname**: name of reference genome that will be put in the comment of the tsv file, e.g. hg19.
If not given it will detect hg18 or hg19 from the comments in the CG file

== Category ==
Format Conversion
