= genome_seq =

== Format ==
cg genome_seq ?options? regionfile dbdir ?outfile?

== Summary ==
Returns sequences of regions in the genome (fasta file), optionally masked for snps/repeats

== Description ==
This command returns the sequences of the genomic regions given in the
file regionfile in fasta format (to stdout or to a file outfile).
Regionfile is a tab delimited file with at least following columns:
chromosome begin end. Repeatmasker repeats are softmasked (lower case) in
the output sequences. Optionally you can hardmask repeats, and soft or
hardmask known (dbsnp) variants based on frequency.

== Arguments ==
; **regionfile**: tab delimited file containing targets with at least following columns: chromosome begin end
; **dbdir**: directory containing reference genomes and variation data

== Options ==
; **-f freq** (--freq): only softmask (lowercase) dbsnp variants if they have a frequency > freq (default is 0, use -1 to include all)
; **-n freqn** (--freqn): only mask (using N) dbsnp variants if they have a frequency > freqn (default is 0.2, use -1 to include all)
; **-d delsize** (--delsize): only mask (using N) dbsnp variants if they are smaller than delsize (default is 5, use -1 to include all)
; **-r repeatmasker** (--repeatmasker): how to mask repeatmasker repeats: "s" means softmask (lowercase), use "N" to mask using Ns, and 0 for no repeatmasking (default is "s")
; **-i idcolumn** (--id): The ids for the fasta file will be taken from the given column (location will be added after a space)
; **-g windowsize** (--gc): add gc content on id line. if windowsize 0 only total gc content will be added. For windowsize > 0, the max gc content for the given windowsize will also be added (default = -1 for no gc content)
; **-c concatseq** (--concat): using this option, all regions will be concatenated into one sequence with concatseq between them. To just concatenate the sequences, use -c ''
; **-m mapfile** (--mapfile): Create a map file that describes which regions in the newly created fasta file map to which regions in the genome
; **--namefield namefield**: entries in the map file will have a name obtained from the **namefield** column in the region file
; **-cn concatname** (--concatname): The concatname wil be the name of sequence in the fasta file generated (if not given, the name will be based on the file)
; **-e concatend** (--concatend): The sequence given by concatend will be added to start and end of the final sequence (only if -c  option was used)
; **-ca concatadj** (--concatadj): The concatseq (-c option) will only be added if regions are separated by at least one base. concatadj will be used to concat adjoining regions (and is '' by default)
; **-gs gccontent** (--gcsplit): Split the result in low and high gc (high has gc >= **gccontent**).
The gc used depends on the -gc option. If -gc is not given, the maxgc at a windowsize of 100 is used.
This option cannot be combined with concatenating sequences, and outfile has to be specified. 
2 files will be generated with lowgc and highgc added in the given outfile name.
== Category ==
Validation
