= Multicompar =

== Format ==
cg multitranscript ?options? multitranscriptfile transcriptfile transcriptfile ?transcriptfile? ...

== Summary ==
Compare multiple transcript files

== Description ==
This command combines multiple transcript files into one multitranscript file; a
tab separated file containing a wide table used to compare transcript (counts)
between different samples. Each line in the table contains general transcript
info (chromosome, begin, end, strand, exonStarts, exonEnds, ?cdsStart?,
?cdsEnd?, ?transcript?, ?gene?, ?geneid?) and columns with transcript info
specific to each sample (count, etc.). The latter have a
column heading of the form field-<sample>.

<sample> can be simply the samplename, but may also include information
about e.g. the sequencing or analysis method in the form method-samplename, e.g.
count_weighed-isoquant-sample1.tsv.

When a transcript was not detected in a sample, the value in the table is set to 0.0

== Arguments ==
; multitranscriptfile: resultfile, will be created if it does not exist
; transcriptfile: file containing transcript of a new sample to be added
  More than one can added in one command

== Options ==
; -match pattern: use approximate matching for (novel) transcripts where the 
  transcript name matches the given pattern (regexp), e.g. "transcript".
  If this is given, transcripts with identical junction chain, but different ends will
  also be matched (the resulting transcript will have the lowest start and highest end position)

== Category ==
Variants
