= Process =

== Format ==
cg process_sample ?options? ?oridir? sampledir

== Summary ==
Processes one sample into a full Genomecomb sample directory ([[projectdir]])

== Description ==

Creates a full Genomecomb sample directory(described in [[projectdir]]) with all
kinds of data, analyses and reports. The directory containing the original
data (fastqs, Complete Genomics data directory) can be given as an option,
an argument, or it can be present in a dir named ori in the sample dir (if
given as an argument, a link named ori to the original data will be made).
The results (which analyses, etc.) differ according to the type of original data.

By default, only files that are not present or that are older than files they
are created from (e.g. because of updates in ori data) will be (re)made.

== Options ==
This command can be distributed on a cluster or using multiple with
[[joboptions|job options]] (more info with cg help joboptions)

As different types of original data are processed differently, not all options
are applicable. Options that are not applicable to the given type of data are 
ignored.

; -dbdir dbdir: Some of the analysis require a reference genome and databases; **dbdir** gives the directory where to find these
; -oridir oridir: directory containing original data, this can be a data directory as it comes from Complete Genomics, or simply a directory containing fastq files or a bam file
; -aligner aligner: use the given aligner for mapping to the reference genome (default bwa)
; -realign value: If **value** is 0, realignment will not be performed, 
use 1 for (default) realignment with gatk, or srma for alignment with srma
; -varcallers varcallers: (space separated) list of variant callers to be used (default "gatk sam")
; -split 1/0: split multiple alternative genotypes over different line
; -paired 1/0: sequenced are paired/unpaired
; -adapterfile file: Use **file** for possible adapter sequences
; -reports list: use all (default) for creating all reports. If you only
want some made, give these as a space separated list. Possible reports are:
flagstats fastqc vars hsmetrics covered
; -samBQ number: only for samtools; minimum base quality for a base to be considered (samtools --min-BQ option)

== Arguments ==
; oridir: directory containing original data, this can be a data directory as it comes from Complete Genomics, or simply a directory containing fastq files or a bam file
; sampledir: name of the sample directory to be created (or completed if it already exists)

== Category ==
Conversion
