= Process_multicompar =

== Format ==
cg process_multicompar ?options? projectdir ?dbdir?

== Summary ==
process a sequencing project directory. 
This expects a genomecomb directory with samples already processed and
makes annotated multicompar data

== Description ==
This command runs the multicomparison step of [[cg_process_project]].
As input, the command expects a basic genomecomb project directory with
sequencing data (projectdir) for which the samples have already been processed.

Some of the steps/commands it uses can be used separately as well:
* [[cg_pmulticompar]]: combine variant data of multiple samples
* [[cg_annotate]]: annotate combined variant data
* [[cg_multireg]]: combine sequenced region data of multiple samples

== Arguments ==
; **projectdir**: project directory with illumina data for different
samples, each sample in a sub directory. The proc will search for fastq
files in dir/samplename/fastq/
; **dbdir**: directory containing reference data (genome sequence,
annotation, ...). dbdir can also be given in a projectinfo.tsv file in the
project directory. process_illumina called with the dbdir parameter will
create the projectinfo.tsv file.

== Options ==
; **-split 1/0**: split multiple alternative genotypes over different line
; **-dbdir dbdir**: **dbdir** can also be given as an option (instead of second parameter)
; **-dbfile file**: Use **file** for extra (files in **dbdir** are already used) annotation 
; **-targetsfile file**: Use this option to easily check certain target positions/variants
in the multicompar. The variants in **file** will allways be added in the 
final multicompar file, even if none of the samples is variant (or even sequenced) in it.
; **-m maxopenfiles** (-maxopenfiles): The number of files that a program
can keep open at the same time is limited. pmulticompar will distribute
the subtasks thus, that the number of files open at the same time stays
below this number. With this option, the maximum number of open files can
be set manually (if the program e.g. does not deduce the proper limit, or
you want to affect the distribution).
; **-varfiles files**: With this option you can limit the varfiles to be added to the given one.
They should be given as a list of files in this option, so if using wildcards, enclose in quotes.

This command can be distributed on a cluster or using multiple with [[joboptions|job options]] (more info with cg help joboptions)

== Category ==
Process
