= Process_illumina =

== Format ==
cg process_illumina ?options? dir ?dbdir?

== Summary ==
process an illumina sequencing project, results in genomecomb dir with multicompar, etc.
By default alignment is done using bwa, duplicates removed (using picard), realigned (using gatk).
Variants are called using gatk and samtools.

== Description ==

== Arguments ==
; **dir**: project directory with illumina data for different samples, each sample in a sub directory.
The proc will search for fastq files in dir/samplename/fastq/
; **dbdir**: directory containing reference data (genome sequence, annotation, ...).
dbdir can also be given in a projectinfo.tsv file in the project directory. process_illumina 
called with the dbdir parameter will create the projectinfo.tsv file.

== Options ==
supports job options (-d)

== Category ==
Conversion
