<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>/home/peter/dev/genomecomb/lib/cg_multigene</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">

<h1 id="h1">Multigene</h1>
<h2 id="h17">Format</h2>
<p>cg multigene ?options? multigenefile genefile genefile ?genefile?
...</p>
<h2 id="h101">Summary</h2>
<p>Compare multiple gene files</p>
<h2 id="h144">Description</h2>
<p>This command combines multiple gene files into one multigene file;
a tab separated file containing a wide table used to compare gene
(counts) between different samples. Each line in the table contains
general gene info (chromosome, begin, end, strand, ?gene?, ?geneid?)
and columns with transcript info specific to each sample (count,
etc.). The latter have a column heading of the form
field-&lt;sample&gt;.</p>
<p>&lt;sample&gt; can be simply the samplename, but may also include
information about e.g. the sequencing or analysis method in the form
method-samplename, e.g. count-isoquant-sample1.</p>
<p>When a gene was not detected in a sample, the value in the table
is set to 0.0.</p>
<p>By default multigene will match based on the gene name (or
geneid), For novel samples (identified based on the presence of a
pattern in the gene name, default containing &quot;novel&quot;)
approximate matching is used: Novel genes will be matched if they
overlap. The resulting gene will have the lowest start and highest
end position of these matches. The id of these result genes will be
changed to one unique for that gene and consistent over different
experiments (based on location and strand)</p>
<h2 id="h1317">Arguments</h2>

<dl>
  <dt>multigenefile</dt>
  <dd>resultfile, will be created if it does not exist</dd>
  <dt>genefile</dt>
  <dd>file containing gene (counts) of a new sample to be added
  More than one can added in one command</dd>
</dl>
<h2 id="h1511">Options</h2>

<dl>
  <dt>-novelpattern pattern</dt>
  <dd>For all genes whose id/name matches <b>pattern</b> (default:
  novel) approximate matching will be used.</dd>
</dl>
<h2 id="h1651">Category</h2>
<p>RNA</p>
</div>

<div id="right">
</div>

</body>
</html>

