<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>/home/peter/dev/genomecomb/lib/cg_multitranscript</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">

<h1 id="h1">Multitranscript</h1>
<h2 id="h23">Format</h2>
<p>cg multitranscript ?options? multitranscriptfile transcriptfile
transcriptfile ?transcriptfile? ...</p>
<h2 id="h137">Summary</h2>
<p>Compare multiple transcript files</p>
<h2 id="h186">Description</h2>
<p>This command combines multiple transcript files into one
multitranscript file; a tab separated file containing a wide table
used to compare transcript (counts) between different samples. Each
line in the table contains general transcript info (chromosome,
begin, end, strand, exonStarts, exonEnds, ?cdsStart?, ?cdsEnd?,
?transcript?, ?gene?, ?geneid?) and columns with transcript info
specific to each sample (count, etc.). The latter have a column
heading of the form field-&lt;sample&gt;.</p>
<p>&lt;sample&gt; can be simply the samplename, but may also include
information about e.g. the sequencing or analysis method in the form
method-samplename, e.g. count_weighed-isoquant-sample1.tsv.</p>
<p>When a transcript was not detected in a sample, the value in the
table is set to 0.0.</p>
<p>By default multitranscript will use approximate matching for novel
transcripts (based on the content of the field &quot;category&quot;):
Novel transcripts will be matched if they have identical junction
chains, even if the (predicted) ends differ. The resulting transcript
will have the lowest start and highest end position of these matches.
The id of these result transcripts will be changed to one unique for
that transcript and consistent over different experiments (based on
location and intron/exon sizes)</p>
<h2 id="h1472">Arguments</h2>

<dl>
  <dt>multitranscriptfile</dt>
  <dd>resultfile, will be created if it does not exist</dd>
  <dt>transcriptfile</dt>
  <dd>file containing transcript of a new sample to be added More
  than one can added in one command</dd>
</dl>
<h2 id="h1675">Options</h2>

<dl>
  <dt>-exact 0/1</dt>
  <dd>use exact matching</dd>
  <dt>-match pattern</dt>
  <dd>For all transcripts whose id/name matches <b>pattern</b>
  (using regexp) approximate matching will be used. (overruling the
  value in category, though this will still be used to see which name
  will be given to the result; known preferred)</dd>
  <dt>-skipempty 0/1</dt>
  <dd>skip isoform files which are empty (size = 0) (default 1)</dd>
</dl>
<h2 id="h2051">Category</h2>
<p>RNA</p>
</div>

<div id="right">
</div>

</body>
</html>

