<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>cg_multicompar_reannot</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">

<h1 id="h1">Multicompar_reannot</h1>
<h2 id="h27">Format</h2>
<p>cg multicompar_reannot ?options? compar_file ?force? ?regonly?
?skipincomplete?</p>
<h2 id="h121">Summary</h2>
<p>Reannotate multicompar file</p>
<h2 id="h164">Description</h2>
<p>cg multicompar-reannot will try to add missing information in a
multicompar file (see cg multicompar), such as whether a variant is
reference or missing (unsequenced) in a sample that did not contain
the variant. To do this, it needs extra information beyond the
variant file. It will try to find this information in several
locations. It will look for files with this information in sample
directories that are siblings to or one dir below the multicompar
file. The file sreg-&lt;source&gt;.tsv should minimally be available
for reannotation (sequenced or not); if not, an error is given.</p>
<p>Some region files can be used to fix missing information</p>

<ul>
  <li>sreg-&lt;source&gt;.tsv should contain all regions that are
  considered to be sequenced. If the variant is located in such a
  region, the sequenced-source value will be changed from ? to r (for
  reference). Outside of these regions it will be annotated as u
  (unsequenced).</li>
  <li>reg_cluster-&lt;source&gt;.tsv: contains regions of clustered
  variants. Used to fill in the cluster column if present</li>
  <li>reg_refcons-&lt;source&gt;.tsv or
  reg_nocall-&lt;source&gt;.tsv: contains poorly called regions
  (complete genomics). Used to fill in the refcons column if present.</li>
</ul>
<p>Coverage (and other) information can be obtained from several
sources (if found)</p>

<ul>
  <li>coverage/coverage-*-&lt;chromosome&gt;.bcol: a binary format
  containing the coverage data for the entire genome (organised per
  chromosome). This format can be created using genomecomb tools, e.g
  from bam files using cg bam2coverage</li>
  <li>varall-&lt;source&gt;.tsv: a tsv file made by returning
  &quot;variant&quot; calling results for all positions. This will
  normally contain the same columns as the variant file.</li>
  <li>coverage/coverageRefScore-&lt;chromosome&gt;-*.tsv (Complete
  Genomics): files provided in a Complete Genomics run containing the
  coverage and refscore for each position of the genome.</li>
</ul>
<h2 id="h2016">Arguments</h2>

<dl>
  <dt><b>compar_file</b></dt>
  <dd>multicompar file</dd>
</dl>
<h2 id="h2069">Options</h2>

<dl>
  <dt><b>-force 0/1</b></dt>
  <dd>if 1, all data will be updated, otherwise only unfilled
  (indicated with a ?) cells will be updated</dd>
  <dt><b>-regonly 0/1</b></dt>
  <dd>if the regonly option is used, only region data (sequenced)
  will be updated</dd>
  <dt><b>-skipincomplete 0/1</b></dt>
  <dd>go on even if data for some samples is not complete (skipping
  samples which are e.g. missing coverage data for some of the
  chromosomes)</dd>
  <dt><b>-paged pagesize</b></dt>
  <dd>reannot needs to keep a lot of files open, too many for
  multicompar files with a lot of samples. Using this option
  reannotation is done per group of <b>pagesize</b> samples. New
  files have to be written for each group</dd>
  <dt><b>-pagedstart start</b></dt>
  <dd>Start paging from a given number (starts with 0). This is
  useful if a -paged run was interupted, you can restart from where
  it stopped.</dd>
</dl>
<p>Some options can also be given as words after the argument</p>

<dl>
  <dt><b>force</b></dt>
  <dd>if present, all data will be updated, otherwise only unfilled
  (indicated with a ?) cells will be updated</dd>
  <dt><b>regonly</b></dt>
  <dd>if the regonly option is used, only region data (sequenced)
  will be updated</dd>
  <dt><b>skipincomplete</b></dt>
  <dd>go on even if data for some samples is not complete (skipping
  samples which are e.g. missing coverage data for some of the
  chromosomes)</dd>
</dl>
<h2 id="h3285">Category</h2>
<p>Variants</p>
</div>

<div id="right">
</div>

</body>
</html>

