<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>cg_makesequenom</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">

<h1 id="h1">Makesequenom</h1>
<h2 id="h20">Format</h2>
<p>cg makesequenom ?options? targetsfile resultfile dbdir</p>
<h2 id="h89">Summary</h2>
<p>Make input files for designing sequenom validation experiments</p>
<h2 id="h167">Description</h2>
<p>This command makes an input file that can be used for sequenom
assay design for the variants in targetsfile. Targetsfile is a tab
delimited file with at least following columns: chromosome begin end
ref alt</p>
<h2 id="h393">Arguments</h2>

<dl>
  <dt><b>targetsfile</b></dt>
  <dd>tab delimited file containing targets with at least following
  columns: chromosome begin end ref alt</dd>
  <dt><b>resultfile</b></dt>
  <dd>sequenom assay setup results</dd>
  <dt><b>dbdir</b></dt>
  <dd>directory containing reference genomes and variation data</dd>
</dl>
<h2 id="h647">Options</h2>

<dl>
  <dt><b>-f freq</b></dt>
  <dd>only softmask (lowercase) dbsnp variants if they have a
  frequency &gt; freq (default is 0, use -1 to include all)</dd>
  <dt><b>-n freq</b></dt>
  <dd>only mask (using N) dbsnp variants if they have a frequency
  &gt; freq (default is 0.2, use -1 to include all)</dd>
  <dt><b>-p snpdbpattern</b></dt>
  <dd>determines which variant databases are used
  (dbdir/var_*snpdbpattern*.tsv.gz). default is &quot;snp&quot; for
  dbsnp. you can e.g. use &quot;Common&quot; for the common variants
  in dbsnp</dd>
  <dt><b>-d delsize</b></dt>
  <dd>only mask (using N) dbsnp variants if they are smaller than
  delsize (default is 5, use -1 to include all)</dd>
  <dt><b>-r 0/s/N</b></dt>
  <dd>if 0, repeatmasker repeats are not masked, N will hardmask
  (replace by N) them. Default is to softmask (s).</dd>
</dl>
<h2 id="h1347">Category</h2>
<p>Validation</p>
</div>

<div id="right">
</div>

</body>
</html>

