<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>/home/peter/dev/genomecomb/lib/cg_multicount</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">

<h1 id="h1">Multicount</h1>
<h2 id="h18">Format</h2>
<p>cg multicount ?options? multicountfile countfile countfile
?countfile? ...</p>
<h2 id="h107">Summary</h2>
<p>Compare multiple count files</p>
<h2 id="h151">Description</h2>
<p>This command combines multiple count files into one multicount
file; a tab separated file containing a wide table used to compare
genes/exon/.. (counts) between different samples. Each line in the
table contains gene/exon/.. names and columns with the count info
specific to each sample (count, etc.). The latter have a column
heading of the form field-&lt;sample&gt;.</p>
<p>&lt;sample&gt; can be simply the samplename, but may also include
information about e.g. the sequencing or analysis method in the form
method-samplename, e.g. count_weighed-isoquant-sample1.tsv.</p>
<p>When a transcript was not detected in a sample, the value in the
table is set to 0.0</p>
<h2 id="h810">Arguments</h2>

<dl>
  <dt>multicountfile</dt>
  <dd>resultfile, will be created if it does not exist</dd>
  <dt>countfile</dt>
  <dd>file containing count data of a new sample to be added More
  than one can added in one command</dd>
</dl>
<h2 id="h1003">Options</h2>

<dl>
  <dt>-idfields list</dt>
  <dd>list with fields to be used as id (default: geneid genename
  gene exon) if present in the count files.</dd>
</dl>
<h2 id="h1138">Category</h2>
<p>Variants</p>
</div>

<div id="right">
</div>

</body>
</html>

