<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>cg_process_project</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">

<h1 id="h1">Process_project</h1>
<h2 id="h23">Format</h2>
<p>cg process_project ?options? projectdir ?dbdir?</p>
<h2 id="h85">Summary</h2>
<p>process a sequencing project directory (<a
href="projectdir.html">projectdir</a>), generating full analysis
information (variant calls, multicompar, reports, ...) starting from
raw sample data from various sources.</p>
<h2 id="h289">Description</h2>
<p>The <a href="cg_process_project.html">cg process_project</a>
command performs the entire secondary analysis (clipping, alignment,
variant calling, reports, ...) and part of the tertiary analysis
(combining samples, annotation, ...) on a number of samples that may
come from various sources. You can specify which analyses should be
run using the options. The default settings/options are for analysing
Illumina genomic sequencing data. You can use the -preset option to
specify default settings for various types of analyses (short or
long-read genome, transcriptome, etc.) A practical example of the
workflow can be found in <a
href="howto_process_project.html">howto_process_project</a>.</p>
<p>The command expects a basic genomecomb project directory (as
described extensively in <a href="projectdir.html">projectdir</a>)
containing a number of samples with raw data (fastq, Complete
genomics results, ...). Each sample is in a separate subdirectory of
a directory named <b>samples</b> in the projectdir. You can add
samples manually or using the <a href="cg_project_addsample.html">cg
project_addsample</a> command as described in <a
href="howto_process_project.html">howto_process_project</a>.</p>
<h3 id="h1371">Per sample analysis</h3>
<p>In the first step, each sampledir is processed using <a
href="cg_process_sample.html">cg process_sample</a>; Samples in one
project can come from different sources (Complete genomics, illumina
sequencing) and be of different types (shotgun, amplicon). Some
options are applied to all samples, e.g. the -amplicons option (for
amplicon sequencing analysis) will place (a link to) the given
amplicons file in each sampledir. These options should only be used
in projects with uniform samples. For mixed samples, these options
can be applied specifically by placing files, e.g. an amplicon file
(named reg_*_amplicons.tsv) in the appropriate sample directories.
More information on specific sample types and options can be found in
the description of <a href="cg_process_sample.html">cg
process_sample</a>.</p>
<h3 id="h2173">Combined analysis</h3>
<p>In the final step process_project will call <a
href="cg_process_multicompar.html">cg process_multicompar</a> to
combine sample results in the subdirectory compar. Different result
files may be present depending on the type of analysis:
<dl>
  <dt>annot_compar-projectname.tsv</dt>
  <dd>multicompar file containing information for all variants in
  all samples (and all methods). If a variant is not present in one
  of the samples, the information at the position of the variant will
  be completed (is the position sequenced or not, coverage, ...) The
  file is also annotated with all databases in dbdir (impact on
  genes, regions of interest, known variant data)</dd>
  <dt>sreg-projectname.tsv</dt>
  <dd>sequenced region multicompar file containing for all regions
  whether they are sequenced (1) or nor (0) for each sample.</dd>
  <dt>annot_cgsv-projectname.tsv</dt>
  <dd>combined results of Complete Genomics structural variant
  calling</dd>
  <dt>annot_cgcnv-projectname.tsv</dt>
  <dd>combined results of Complete Genomics CNV calling</dd>
</dl>
<h2 id="h3143">Arguments</h2>

<dl>
  <dt><b>projectdir</b></dt>
  <dd>project directory with illumina data for different samples,
  each sample in a sub directory. The proc will search for fastq
  files in dir/samplename/fastq/</dd>
  <dt><b>dbdir</b></dt>
  <dd>directory containing reference data (genome sequence,
  annotation, ...). dbdir can also be given in a projectinfo.tsv file
  in the project directory. process_project called with the dbdir
  parameter will create the projectinfo.tsv file.</dd>
</dl>
<h2 id="h3579">Options</h2>
<p>This command can be distributed on a cluster or using multiple
with <a href="joboptions.html">job options</a> (more info with cg
help joboptions)</p>
<p>As different types of original data are processed differently, not
all options are applicable. Options that are not applicable to the
given type of data are ignored.</p>
<p>By default options are set for short read genomic sequencing
(genome/exome/targeted). Presets can be used to set a number of
options to the defaults for a given analysis. These options can still
be changed by specifically giving them a value after the -preset
option (options given later overrule previous ones).</p>

<dl>
  <dt>-preset preset</dt>
  <dd>sets a number of options to the defaults for the given
  &quot;preset&quot;, must be one of: srs (short read genomic
  sequencing), rseq (short read rna-seq), ont (ont genomic
  sequencing), ontr (ont RNA-seq), scywalker (ont 10x single cell
  rna-seq), scywalker_pacbio (pacbio 10x single cell rna-seq)</dd>
  <dt>-samplesheet samplesheet</dt>
  <dd>create <b>projectdir</b> based on the data in the given
  <b>sampleheet</b> as described in <a href="cg_make_project.html">cg
  make_project</a></dd>
  <dt>-dbdir dbdir</dt>
  <dd><b>dbdir</b> can also be given as an option (instead of
  second parameter)</dd>
  <dt>-minfastqreads num</dt>
  <dd>fastq based samples with less than <b>num</b> reads in the
  fastq files are not processed and not added to the final compar.</dd>
  <dt>-clip</dt>
  <dd>clip adaptor sequences prior to alignment using fastq-mcf
  (default 1)</dd>
  <dt>-paired 1/0 (-p)</dt>
  <dd>sequenced are paired/unpaired</dd>
  <dt>-adapterfile file</dt>
  <dd>Use <b>file</b> for possible adapter sequences</dd>
  <dt>-removeskew num</dt>
  <dd>-k parameter for sequence clipping using fastq-mcf: sKew
  percentage-less-than causing cycle removal</dd>
  <dt>-aligners aligner (-a)</dt>
  <dd>use the given aligner for mapping to the reference genome
  (default bwa) Currently supported are: bwa, bowtie2, minimap2_sr,
  minimap2, minimap2_pb, minimap2_asm20, ngmlr ; for rna-seq: star,
  star_2p, hisat2, minimap2_splice, minimap2_splicehq</dd>
  <dt>-ali_keepcomments 1/0</dt>
  <dd>set to 1 to transfer sequence comments in the source fastq or
  ubams to the alignment (default don't keep for fastq, keep for
  ubams). This option currently only works for minimap2 aligner</dd>
  <dt>-aliformat format</dt>
  <dd>format of the (final) alignment (map) files, this is by
  default bam, but can be set to cram</dd>
  <dt>-realign value</dt>
  <dd>If <b>value</b> is 0, realignment will not be performed, use
  1 (default) for realignment with gatk, or <b>value</b> srma for
  alignment with srma</dd>
  <dt>-removeduplicates 0/1/picard/biobambam</dt>
  <dd>By default duplicates will be removed (marked actually) using
  samtools except for amplicon sequencing. With this option you can
  specifically request or turn of duplicate removal (overruling the
  default). If you want to use large amounts of memory ;-), you can
  still use picard for removing duplicates (third option)</dd>
  <dt>-amplicons ampliconfile</dt>
  <dd>This option turns on amplicon sequencing analysis (as
  described in <a href="cg_process_sample.html">cg
  process_sample</a>) using the amplicons defained in
  <b>ampliconfile</b> for all samples that do not have a sample
  specific amplicon file yet.</dd>
  <dt>-varcallers varcallers</dt>
  <dd>(space separated) list of variant callers to be used (default
  &quot;gatkh strelka&quot;). Currently supported are: gatk, gatkh
  (haplotype caller), strelka, sam, freebayes, bcf, longshot, clair3,</dd>
  <dt>-svcallers svcallers</dt>
  <dd>(space separated) list of structural variant callers to be
  used (default empty). Currently supported are: manta, lumpy, gridds
  sniffles, cuteSV, npinv</dd>
  <dt>-methcallers methcallers</dt>
  <dd>(space separated) list of methylation callers to be used
  (default empty). Currently supported are: nanopolish</dd>
  <dt>-reftranscripts reftranscripts</dt>
  <dd>file with reference transcripts for isoform calling. (default
  empty -&gt; finds default in refdb) Currently supported are: flair,
  isoquant, flames</dd>
  <dt>-isocallers isocallers</dt>
  <dd>(space separated) list of isofrom calling (and counting)
  programs to be used for rna-seq data. Currently supported are:
  flair, isoquant, flames</dd>
  <dt>-iso_joint isocallers</dt>
  <dd>perform joint analysis on the given list of isoform callers:
  the resulting isoforms of all per sample runs using the isoform
  caller (so must be in -isocallers) are combined, and each sample is
  recalled with the combination of isoforms (including novel ones
  found in at least 2 samples default) as reference</dd>
  <dt>-iso_joint_min integer</dt>
  <dd>minimum number of samples an isoform has to be found in to be
  included in the joint analysis</dd>
  <dt>-organelles organelles</dt>
  <dd>(space separated) list of chromosomes that are organelles
  (that are treated differently in some analysis) If not given
  explicitely, the ones indicated in the file
  $refdb/extra/reg_*_organelles.tsv (if present) will be used</dd>
  <dt>-counters counters</dt>
  <dd>(space separated) list of counter programs to be used for
  rna-seq data. Currently supported are: rnaseqc, qorts</dd>
  <dt>-split 1/0</dt>
  <dd>split multiple alternative genotypes over different line</dd>
  <dt>-downsampling_type NONE/ALL_READS/BY_SAMPLE/</dt>
  <dd>sets the downsampling type used by GATK (empty for default).</dd>
  <dt>-reports list</dt>
  <dd>use basic (default) for creating most reports, or all for all
  reports. If you only want some made, give these as a space
  separated list. Possible reports are (further explained in <a
  href="cg_process_reports.html">cg process_reports</a>): fastqstats
  fastqc flagstat_reads flagstat_alignments samstats alignedsamstats
  unalignedsamstats histodepth vars hsmetrics covered histo
  predictgender singlecell</dd>
  <dt>-m maxopenfiles (-maxopenfiles)</dt>
  <dd>The number of files that a program can keep open at the same
  time is limited. pmulticompar will distribute the subtasks thus,
  that the number of files open at the same time stays below this
  number. With this option, the maximum number of open files can be
  set manually (if the program e.g. does not deduce the proper limit,
  or you want to affect the distribution).</dd>
  <dt>-samBQ number</dt>
  <dd>only for samtools; minimum base quality for a base to be
  considered (samtools --min-BQ option)</dd>
  <dt>-distrreg regions</dt>
  <dd>distribute regions for parallel processing. Possible options
  are <b> 0: no distribution (also empty) </b> 1: default
  distribution <b> schr or schromosome: each chromosome processed
  separately </b> chr or chromosome: each chromosome processed
  separately, except the unsorted, etc. with a _ in the name that
  will be combined), <b> a number: distribution into regions of this
  size </b> a number preceded by an s: distribution into regions
  targeting the given size, but breaks can only occur in unsequenced
  regions of the genome (N stretches) <b> a number preceded by an r:
  distribution into regions targeting the given size, but breaks can
  only occur in large (&gt;=100000 bases) repeat regions </b> a
  number preceded by an g: distribution into regions targeting the
  given size, but breaks can only occur in large (&gt;=200000 bases)
  regions without known genes <b> g: distribution into regions as
  given in the &lt;refdir&gt;/extra/reg_*_distrg.tsv file; if this
  does not exist uses g5000000 </b> a file name: the regions in the
  file will be used for distribution</dd>
  <dt>-maxfastqdistr maxfastqdistr</dt>
  <dd>if there are more than <b>maxfastqdistr</b> separate input
  fastqs, they will be merged into <b>maxfastqdistr</b> fastqs for
  analysis: If there are many (small) fastqs, the overhead to
  processes (alignment etc.) them separately (default, to distribute
  the load) can become too large.</dd>
  <dt>-datatype datatype</dt>
  <dd>Some variant callers (strelka) need to know the type of data
  (genome, exome or amplicons) for analysis. You can specify it using
  this option. If not given, it is deduced from acompanying region
  files (reg_*_amplicons.tsv for ampicons or reg_*_amplicons.tsv for
  exome)</dd>
  <dt>-hap_bam 0/1</dt>
  <dd>if 1 produce a bam file with haplotype indictions (longshot
  only) (default 0)</dd>
  <dt>-depth_histo_max number</dt>
  <dd>in reports, count positions with up to <b>number</b> depth
  (default 1000). Larger dfepths will be counted under <b>number</b></dd>
  <dt>-targetfile targetfile</dt>
  <dd>if targetfile is provided, coverage statistics will be
  calculated for this region</dd>
  <dt>-targetvarsfile file</dt>
  <dd>Use this option to easily check certain target
  positions/variants in the multicompar. The variants in <b>file</b>
  will allways be added in the final multicompar file, even if none
  of the samples is variant (or even sequenced) in it.</dd>
  <dt>-dbfile file</dt>
  <dd>Use the given <b>file</b> for extra (files in <b>dbdir</b>
  are already used) annotation. This option can be given more than
  once; all given files will be added</dd>
  <dt>-dbfiles files</dt>
  <dd>Use <b>files</b> for extra (files in <b>dbdir</b> are already
  used) annotation. <b>files</b> should be a space separated list of
  files.</dd>
  <dt>-conv_nextseq 1/0</dt>
  <dd>generate fastqs for nextseq run &amp; create sample folders -
  rundir should be placed in projectdir of resulting variants. This
  option can be added multiple times (with different files)</dd>
  <dt>-jobsample integer</dt>
  <dd>By default (0) the processing of each sample is split in many
  separate jobs. If you have to process many samples with relatively
  short individual runtimes or your cluster limits the number of jobs
  you can set this to 1 or more to run each sample in only one job,
  thus reducing the job managment overhead. The number given is the
  number of cores assigned to each such job.</dd>
  <dt>-keepfields fieldlist</dt>
  <dd>Besides the obligatory fields, include only the fields in
  fieldlist (space separated) in the multicompar file. Default is to
  use all fields present in the file (*). All fields will still be
  used in the per sample output.</dd>
</dl>
<p>Following options are available for singlecell or bulk UMI
corrected analysis:
<dl>
  <dt>-singlecell o/1/ontr10x</dt>
  <dd>set to 1 or ontr10x for (scywalker) long read single cell
  analysis, only supported for isoquant isoform caller</dd>
  <dd>-sc_expectedcells number: gives the the number of cells
  expected.</dd>
  <dd>-cellmarkerfile file: A tsv file providing genes that are
  indicative specific cell cell types.</dd>
  <dt>-tissue string</dt>
  <dd>The tissue type of the sample. If cellmarkerfile is given,
  only markers of the given tissue are used.</dd>
  <dt>-sc_whitelist file</dt>
  <dd>(for singlecell) file with all possible correct barcodes. By
  default the whitelist with the 10x version 3 barcodes is used. You
  can specify to use version 2 of the whitelist by using the shortcut
  v2</dd>
  <dt>-sc_adaptorseq string</dt>
  <dd>adapter sequence used to find the barcode and UMI. These
  should follow the given sequence in the reads (default is
  CTACACGACGCTCTTCCGATCT for singlecell)</dd>
  <dt>-sc_barcodesize number</dt>
  <dd>size of the barcode (default 16 for singlecell)</dd>
  <dt>-sc_umisize number</dt>
  <dd>size of the UMI (default 12 for singlecell)</dd>
  <dt>-addumis 0/1</dt>
  <dd>when 1, add barcode and UMI info to reads for bulk
  sequencing, in (bulk) isoquant analysis umicounts will be added in
  the results (barcodes are not used) You can use -sc_adaptorseq,
  -sc_barcodesize and -sc_umisize to change how barcodes and UMIs are
  searched (default for -addumis is
  GATCGGAAGAGCACACGTCTGAACTCCAGTCAC,8,12)</dd>
</dl>
<p>This command can be distributed on a cluster or using multiple
cores with <a href="joboptions.html">job options</a> (more info with
cg help joboptions) The option -distrreg can be used to allow a
greater distribution by doing some analyses (variantcalling,
annotations) split by region (chromosomes) and combining the results</p>
<h2 id="h14496">Sample specific options</h2>
<p>Different options can be given to different samples within the
same experiment run by storing a file named <b>options.tsv</b> in the
experiment/project dir with the following fields: sample option value</p>
<p>For each sample (that differs from the general option if given)
you add a line with the samplename, the option (without the -, e.g.
sc_expectedcells) and the value (the number of expected cells in the
case of sc_expectedcells). Sample specific options given this way
overrule the general options given on the process_project commandline
(for that sample)</p>
<p>You can also use the preset option this way, allowing the analysis
of different technologies (e.g. ont and srs) in one run. Beware that
presets just change base/default settings; Options explicitely given
on the process_project commandline will overrule settings from a
preset, even if sample specific).</p>
<h2 id="h15395">Dependencies</h2>
<p>Some of the programs used in this workflow are not distributed
with genomecomb itself (e.g. gatk, strelka) and should be installed
separately. To make this easier, they are available as portable
application directories from the [genomecomb
website](https://derijkp.github.io/genomecomb/install.html) or can be
directly installed using the <a href="cg_install.html">cg install</a>
command.</p>
<h2 id="h15789">Example</h2>
<pre>
cg process_project -d sge testproject /complgen/refseq/hg19
</pre>
<h2 id="h15872">Category</h2>
<p>Process</p>
</div>

<div id="right">
</div>

</body>
</html>

