<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>/home/peter/dev/genomecomb/lib/cg_sv</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">

<h1 id="h1">Var</h1>
<h2 id="h11">Format</h2>
<p>cg sv ?options? bamfile ?resultfile?</p>
<h2 id="h62">Summary</h2>
<p>generic command to call structural variants based on the alignment
in bamfile.</p>
<h2 id="h156">Description</h2>
<p>The command can be used to call structural variants using
different methods. The methods are implemented in separate commands
such as cg sv_manta; The -method options determines which one is
used. The <b>resultfile</b> is a tsv file containing the called
structural variants. If &quot;&quot;resultfile<b> is not given, a
file name will be derived from the bam file name: If the bamfile has
a name of the form map-alignmethod-sample.bam, the result file will
will be var-svcaller-alignmethod-sample.tsv.zst. Other result files
may also be created, depending on the method used.</b></p>
<p>cg sv can distribute structural variant calling by analysing
separate regions in parallel and combining the results (-distrreg
option). This command can be distributed on a cluster or using
multiple cores with <a href="joboptions.html">job options</a> (more
info with cg help joboptions)</p>
<h2 id="h1012">Arguments</h2>

<dl>
  <dt>bamfile</dt>
  <dd>alignment on which to call variants</dd>
  <dt>resultfile</dt>
  <dd>filename of the resulting structural variant calls</dd>
</dl>
<h2 id="h1141">Options</h2>

<dl>
  <dt>-method svcaller</dt>
  <dd>use the given method (<b>svcaller</b>) to call structural
  variants. Default svcaller is manta. Possible values are: manta,
  lumpy sniffles, npinv, cg, gridds</dd>
  <dt>-refseq refseq</dt>
  <dd>genomic reference sequence</dd>
  <dt>-regionfile regionfile</dt>
  <dd>only call variants in the region(s) given in regionfile</dd>
  <dt>-regmincoverage mincoverage</dt>
  <dd>If no regionfile is given, variants will be called only in
  regions with a coverage of at least <b>mincoverage</b> (default 3)</dd>
  <dt>-distrreg</dt>
  <dd>distribute regions for parallel processing (default
  s50000000). Possible options are <b> 0: no distribution (also
  empty) </b> 1: default distribution (s50000000) <b> schr or
  schromosome: each chromosome processed separately </b> chr or
  chromosome: each chromosome processed separately, except the
  unsorted, etc. with a _ in the name that will be combined), <b> a
  number: distribution into regions of this size </b> a number
  preceded by an s: distribution into regions targeting the given
  size, but breaks can only occur in unsequenced regions of the
  genome (N stretches) <b> a number preceded by an r: distribution
  into regions targeting the given size, but breaks can only occur in
  large (&gt;=100000 bases) repeat regions </b> a number preceded by
  an g: distribution into regions targeting the given size, but
  breaks can only occur in large (&gt;=200000 bases) regions without
  known genes <b> a filename: the regions in the file will be used
  for distribution</b></dd>
  <dt>-split 1/0</dt>
  <dd>indicate if in the datafiles, multiple alternative genotypes
  are split over different lines (default 1)</dd>
  <dt>-t number (-threads)</dt>
  <dd>number of threads used by each job (if the method supports
  threads)</dd>
  <dt>-cleanup 0/1</dt>
  <dd>Use 0 to no cleanup some of the temporary files (vcf, delvar,
  ...)</dd>
</dl>
<h2 id="h2860">Category</h2>
<p>Variants</p>
</div>

<div id="right">
</div>

</body>
</html>

