<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>/home/peter/dev/genomecomb/lib/cg_tsv2vcf</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">

<h1 id="h1">tsv2vcf</h1>
<h2 id="h15">Format</h2>
<p>cg tsv2vcf ?options? ?infile? ?outfile?</p>
<h2 id="h69">Summary</h2>
<p>Converts data in genomecomb tab-separated variant format (<a
href="tsv.html">tsv</a>) to vcf.</p>
<h2 id="h159">Description</h2>
<p>cg tsv2vcf converts a genomecomb tab-separated variant file (<a
href="tsv.html">tsv</a>) to a vcf. If metadata is present in the tsv
file (in the genomecomb metadata format), it will be used for
creating the vcf header. If the metadata is not present, A number of
default descriptions for the fieldnames are used to create the header
(these are based on the gatk(h), samtools, etc. variantcallers) The
commands needs the genomic reference sequence, which can be given
directly using the -refseq option or indirectly using -dbdir.</p>
<h2 id="h689">Arguments</h2>

<dl>
  <dt>infile</dt>
  <dd>file to be converted, if not given, uses stdin. File may be
  compressed.</dd>
  <dt>outfile</dt>
  <dd>write results to outfile, if not given, uses stdout</dd>
</dl>
<h2 id="h851">Options</h2>

<dl>
  <dt>-split 0/1</dt>
  <dd>tsv files usually have all variants on separate lines. cg
  tsv2vcf will by default merge variants that start at the same
  position (as usual in the vcf format). Using -split 1, will put
  these on separate lines as well</dd>
  <dt>-refseq refseq</dt>
  <dd>genomic reference sequence</dd>
  <dt>-dbdir dbdir</dt>
  <dd>dbdir containing reference genomes databases, used to find
  the genomic reference</dd>
  <dt>-sample samplename</dt>
  <dd>For single sample tsv files (not multicoompar) the samplename
  that will be used in the vcf file can be given using this option.
  If not given, it will try to find the samplename in the file
  (metadata), or base it on the filename</dd>
</dl>
<h2 id="h1489">Category</h2>
<p>Format Conversion</p>
</div>

<div id="right">
</div>

</body>
</html>

