= tsv format =

The standard file format used in GenomeComb is the widely supported,
simple, yet flexible tab-separated values file (tsv). 
A tsv file is a simple text file containing tabular data, where each line
represents a record or row in the table.
Each field value of a record is separated from the next by a tab character.

The first line of the tab file (not starting with a #) is a
header that contains column names. Comment lines (indicated by starting the
line with a # character) may precede the header; They may contain meta
data.
Most Genomecomb commands will detect and support some alternative comments/header styles:
; sam: starts with "@HD	VN", header lines start with @, uses fixed columns
; vcf: starts with "##fileformat=VCF", the last "comment" line contains the header.
In order to extract the data merged in some of the vcf fields into a genomecomb supported tsv,
use [[cg_vcf2tsv|cg vcf2tsv]]
; Complete genomics: header line is preceeded by an empty line and starts with a > character
; 

As the tsv format does not use quoting, values in the table cannot contain tabs or newlines,
unless by coding them e.g. by using escape characters (\t,\n)

== Gnomecomb specific tsv files ==
Depending on the columns present, tsv files can be used for various purposes. 
Genomecomb gives following columns specific meanings:

; **chromosome** or **chrom**: chromosome name. Many genomecomb tools allow mixing chr1 and 1 notations
; **begin** or **start**: start of feature. half-open coordinates as used by UCSC bed files and
Complete Genomics files are expected. This means for instance that the first base of a sequence will
be indicated by start=0 and end=1. An insertion before the first base will have start-0, end=0.
; **end**: end of feature in half-open coordinates
; **type**: type of variation: snp, ins, del, sub are recognised
; **ref** or **reference**: genotype of the reference sequence at the feature. For large deletions, the size 
of the deletion can be used.
; **alt**: alternative genotype(s). If there are more than one alternatives, they are 
separated by commas.
; **alleleSeq1**: gentype of features at one allele
; **alleleSeq2**: gentype of features at other allele

Most tools expect the files to be sorted on chromosme,begin,end,type and will create sorted files.
You can sort files using the -s option of [[cg_select]].
Not all the columns must be present, and any other columns can be added and searched. In 
files containing data for multiple samples, columns that are specific to a sample have -samplename 
appended to the column name. Some examples of (minimal) columns present for various genomecomb files:

; **region file**: chromosome begin end.
; **variant file**: chromosome begin end type ref alt ?alleleSeq1? ?alleleSeq2?
; **multicompar file**: chromosome begin end type ref alt alleleSeq1-sample1 alleleSeq2-sample1 alleleSeq1-sample2 alleleSeq2-sample2 ...

These files can easily be queried using the [[cg_select]] functionality or can
be loaded into a local database.
