<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>GenomeComb</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">




<h1>GenomeComb</h1>

<h1 class=subtitle>Home</h1>

<h2>Synopsis</h2>
<p>Genomecomb is an integrated, flexible package for the analysis of
genome and transcriptome sequencing, supporting the entire workflow for
both short and long read sequencing data: It provides both a pipeline to
produce files such as annotated variant and structural variant files, and
transcript predictions and counts starting from raw data, as well as tools
to further to analyse these by for example flexible filtering according to
your project.</p>

<h2>Main features</h2>

<p>In addition to providing <a href="install.html" >installation directives</a>, 
extensive <a href="intro.html">documentation</a>, we also provide a shortlist of 
common "<a href="howto.html">protocols</a>" used to analyse whole genome sequencing data. 
A short list of analyses that can be performed using GenomeComb is given below:</p>
<ul>
<li>Flexible analysis pipeline
<ul>
  <li>Starting from raw data, can run combinations of multiple, selectable analysis 
      tools in parallel and integrate the results into easily searchable files. 
  <li>Supports genome as well as trancriptome sequencing using Illumina as well as Nanopore data
  <li>Can run on a single machine (using multiple cores) or on heavily distributed on a cluster.
  <li>Extensive annotation, using publicly available data (e.g. UCSC tracks) or custom annotation tracks
</ul>
<li>Powerful queries
<ul>
  <li>All files (variant, genome region, transcipt, ..) files can be filtered using advanced queries,
      supporting combinations of multiple analysis methods
  <li>Both a command line and a graphical interface
  <li>Supports browsing and querying the large data (several gigabytes) generated in NGS.
  <li>Adding calculated fields to a file
  <li>Summaries
  <li>Results and summaries can also be further queried
</ul>
<li>Other tools
<ul>
  <li>Genetic analyses (homwes, compound heterozygotes, ...)
  <li>Quantification of the sequenced and filtered regions in a genome
  <li>Set operations on genome regions and variant files: compare, join, subtract
  <li>Automated primer design for Sanger or Sequenom validation directly from the variant files
</ul>
</ul>

<h2>News</h2>
<ul>
</ul>

<h2>Citation</h2>

<p>We have used a very early version of GenomeComb in the analysis of whole genome sequences of monozygotic twin genomes, 
tumor-normal genomes and publicly available HapMap genomes.
A citable publication about this study has appeared in Nature Biotch:</p>
<p>
<i>
Reumers, J*, De Rijk, P*, Zhao, H, Liekens, A, Smeets, D, Cleary, J, Van Loo, P, Van Den Bossche, M, Catthoor, K, Sabbe, B, Despierre, E, Vergote, I, Hilbush, B, Lambrechts, D and Del-Favero, J
Optimized filtering reduces the error rate in detecting genomic variants by short-read sequencing.<br>
Nature biotechnology, 30, 61-88<br>
<a href="http://www.ncbi.nlm.nih.gov/pubmed/22178994" target="_blank">pubmed 22178994</a><br>
</i>
</p>

<p>For the homwes tool available in genomecomb the folowing paper should be referenced
<p><i>
Kancheva,D, Atkinson,D, De Rijk,P, Zimon,M, Chamova,T, Mitev,V, Yaramis,A, Maria Fabrizi,G, Topaloglu,H, Tournev,I, Parma, Y, Battaloglu, E, Estrada-Cuzcano, A, Jordanova, A (2015)
Novel mutations in genes causing hereditary spastic paraplegia and Charcot-Marie-Tooth neuropathy identified by an optimized protocol for homozygosity mapping based on whole-exome sequencing.
Genet. Med., 10.1038<br>
<a href="http://www.ncbi.nlm.nih.gov/pubmed/26492578" target="_blank">pubmed 26492578</a><br>
</i></p>

<h2>License</h2> 
<p>This software is available under the conditions of the <a href="http://www.gnu.org/licenses/gpl.txt">GPL</a></p>



</div>

<div id="right">
</div>

</body>
</html>

