<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>/home/peter/dev/genomecomb/help/intro</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">

<h1 id="h1">Introduction</h1>
<h2 id="h20">GenomeComb</h2>
<p>GenomeComb provides tools to analyze, combine, annotate and query
whole genome, exome or targetted sequencing data. Variant files in <a
href="format_tsv.html">tab-separated format</a> from different
sequencing datasets can be <a
href="howto_process_project.html">generated</a> and <a
href="cg_pmulticompar.html">combined</a> taking into account which
regions are actually sequenced (given as region files in
tab-separated format), <a href="cg_annotate.html">annotated</a> and <a
href="cg_select.html">queried</a> (several examples can be seen in
the <a href="howto_query.html">Howto_Query</a> section). A <a
href="cg_viz.html">graphical user interface</a> able to browse and
query multi-million line tab-separated files is also included.</p>
<p>The <a href="cg_process_project.html">cg process_project</a>
command provides a pipeline to generate annotated multisample variant
data, reports, etc. starting from various raw data source material
(e.g. fastq files,Complete Genomics data), which can be run locally
(optionally using multiple cores) or distributed on a cluster. It
combines many of the genomecomb commands that are also available
separately (<a href="reference.html">reference</a>)</p>
<h2 id="h1082">File formats</h2>
<p>The standard file format used in GenomeComb is the widely
supported, simple, yet flexible tab-separated values file (<a
href="format_tsv.html">format_tsv</a>). This text format contains
tabular data, where each line is a record, and each field is
separated from the next by a TAB character. The first line (not
starting with a #) is a header indicating the names of each column
(or field). Lines starting with a # preceeding the header are
comments (and may store metadata on the file). The file extension
.tsv can be used to refer to this format.</p>
<p>Depending on which columns are present, tsv files can be used for
various purposes. Usually the files are used to describe features on
a reference genome sequence. In this intro a number of basic fields
and uses are described. Refer to the <a
href="format_tsv.html">format_tsv</a> help for more in depth info on
the format and its uses. Some typical fields are:</p>

<dl>
  <dt>chromosome</dt>
  <dd>chromosome name. Many genomecomb tools allow mixing chr1 and
  1 notations</dd>
  <dt>begin</dt>
  <dd>start of feature. half-open coordinates as used by UCSC bed
  files and Complete Genomics files are expected. This means for
  instance that the first base of a sequence will be indicated by
  start=0 and end=1. An insertion before the first base will have
  start-0, end=0.</dd>
  <dt>end</dt>
  <dd>end of feature in half-open coordinates</dd>
  <dt>type</dt>
  <dd>type of variation: snp, ins, del, sub are recognised</dd>
  <dt>ref</dt>
  <dd>genotype of the reference sequence at the feature. For large
  deletions, the size of the deletion can be used.</dd>
  <dt>alt</dt>
  <dd>alternative genotype(s). If there are more than one
  alternatives, they are separated by commas.</dd>
  <dt>alleleSeq1</dt>
  <dd>gentype of features at one allele</dd>
  <dt>alleleSeq2</dt>
  <dd>gentype of features at other allele</dd>
</dl>
<p>Most tools expect the tsv files to be sorted on
chromosome,begin,end,type and will create sorted files. You can sort
files using the -s option of <a href="cg_select.html">cg select</a>.
Not all the columns must be present, and any other columns can be
added and searched. In files containing data for multiple samples,
columns that are specific to a sample have -samplename appended to
the column name. Some examples of (minimal) columns present for
various genomecomb files:</p>

<dl>
  <dt>region file</dt>
  <dd>chromosome begin end.</dd>
  <dt>variant file</dt>
  <dd>chromosome begin end type ref alt ?alleleSeq1? ?alleleSeq2?</dd>
  <dt>multicompar file</dt>
  <dd>chromosome begin end type ref alt alleleSeq1-sample1
  alleleSeq2-sample1 alleleSeq1-sample2 alleleSeq2-sample2 ...</dd>
</dl>
<p>These files can easily be queried using the <a
href="cg_select.html">cg select</a> functionality or can be loaded
into a local database.</p>
<p>The format does not use quoting, so values in the table cannot
contain tabs or newlines, unless by coding them using escape
characters (\t,\n)</p>
<h2 id="h3732">Project directories</h2>
<p>While not necessary for many of the commands, using the specific
organisation of files in a Genomecomb project directory (described in
<a href="projectdir.html">projectdir</a>) is useful: e.g. the process
commands (e.g. <a href="cg_process_project.html">cg
process_project</a>,<a href="cg_process_sample.html">cg
process_sample</a>) to run an entire analysis pipeline expect this
structure to start from and generates all additional data in this
structure.</p>
<h2 id="h4123">How to start</h2>
<p>In the <a href="howto.html">Howto</a> section we give some
extended examples on how to process ngs data and query the results.</p>
</div>

<div id="right">
</div>

</body>
</html>

