<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd">

<html>
<head>
<title>sam_vars</title>

<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta name="MSSmartTagsPreventParsing" content="TRUE">
<meta name="keywords" content="GenomeComb, genome sequencing, genome comparison, filtering, annotation">
<meta http-equiv="expires" content="-1">
<meta http-equiv= "pragma" CONTENT="no-cache">

<link rel="stylesheet" href="css/default.css" type="text/css"/> 

<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-24207049-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</head>

<body bgcolor="#ffffff" text="#000000">


<iframe class="ahem"><font color="#808080"><big>[You are using an out of date browser. This site will look better if you <a href="http://www.webstandards.org/upgrade/" target="ala" title="The Web Standards Project's BROWSER UPGRADE initiative.">upgrade</a> to a current browser that supports web standards. The upgrade link will take you to the new versions of Netscape, Explorer, and Opera. Thanks.]</big></font></iframe>
<a name="top"> </a>
<div id="top">
<p class="top"><br>GenomeComb</p>
</div>

<div id="left">
<h2 class = "menu_2"><a href="index.html">Home</a></h2>
<h2 class = "menu_2"><a href="contact.html">Contact</a></h2>
<h2 class = "menu_2"><a href="install.html">Installation</a></h2>
<h2 class = "menu_2"><a href="intro.html">Documentation</h2>
<p class = "menu_mt"><a href="intro.html">Introduction</a></p>
<p class = "menu_mt"><a href="reference.html">Reference</a></p>
<p class = "menu_mt"><a href="howto.html">Howtos</a></p>
</div>

<div id="middle">

<h1 id="h1">samtools based variant file</h1>
<p>Variant files created by genomecomb using the samtools variant
caller are in the usual genomecomb tab-separated variant file format
(<a href="tsv.html">tsv</a>). besides the usual fields of a variant
tsv file, they (can) also contain the following columns:</p>

<dl>
  <dt>totalcoverage</dt>
  <dd>Raw read depth (DP in vcf INFO); samtools does not count all
  mapping reads for totalcoverage (bad reads are filtered out)</dd>
  <dt>coverage</dt>
  <dd>number of high-quality bases (DP in vcf FORMAT); bases with
  quality &lt; 13 are filtered out as well</dd>
  <dt>genoqual</dt>
  <dd>genotype quality, encoded as a phred quality
  -10log_10p(genotype call is wrong)</dd>
  <dt>phased</dt>
  <dd>1 if genotype is phased, 0 otherwise</dd>
  <dt>genotypes</dt>
  <dd>list of allele numbers (0 for ref, 1 for first alt, ...) (GT
  in vcf) The alleles are comma (,) separated for phased genotypes (|
  in vcf), and &quot;;&quot; separated for unphased (/ in vcf). In
  contrast to alleleSeq1 and alleleSe2, this field allows for
  ploidies other than 2</dd>
  <dt>loglikelihood</dt>
  <dd>Likelihoods for RR,RA,AA genotypes (R=ref,A=alt)</dd>
  <dt>DV</dt>
  <dd>number of high-quality non-reference bases</dd>
  <dt>SP</dt>
  <dd>Phred-scaled strand bias P-value</dd>
  <dt>PL</dt>
  <dd>List of Phred-scaled genotype likelihoods</dd>
  <dt>DP4</dt>
  <dd>number of high-quality ref-forward bases, ref-reverse,
  alt-forward and alt-reverse bases</dd>
  <dt>MQ</dt>
  <dd>Root-mean-square mapping quality of covering reads</dd>
  <dt>FQ</dt>
  <dd>Phred probability of all samples being the same</dd>
  <dt>AF1</dt>
  <dd>Max-likelihood estimate of the first ALT allele frequency
  (assuming HWE)</dd>
  <dt>AC1</dt>
  <dd>Max-likelihood estimate of the first ALT allele count (no HWE
  assumption)</dd>
  <dt>totalallelecount</dt>
  <dd>Total number of alleles in called genotypes</dd>
  <dt>IS</dt>
  <dd>Maximum number of reads supporting an indel and fraction of
  indel reads</dd>
  <dt>allelecount</dt>
  <dd>Allele count in genotypes for each ALT allele, in the same
  order as listed</dd>
  <dt>G3</dt>
  <dd>ML estimate of genotype frequencies</dd>
  <dt>HWE</dt>
  <dd>Chi^2 based HWE test P-value based on G3</dd>
  <dt>CLR</dt>
  <dd>Log ratio of genotype likelihoods with and without the
  constraint</dd>
  <dt>UGT</dt>
  <dd>The most probable constrained genotype configuration in the
  trio</dd>
  <dt>CGT</dt>
  <dd>P-values for strand bias, baseQ bias, mapQ bias and tail
  distance bias</dd>
  <dt>PV4</dt>
  <dd>P-values for strand bias, baseQ bias, mapQ bias and tail
  distance bias</dd>
  <dt>INDEL</dt>
  <dd>Indicates that the variant is an INDEL.</dd>
  <dt>PC2</dt>
  <dd>Phred probability of the nonRef allele frequency in group1
  samples being larger (,smaller) than in group2.</dd>
  <dt>PCHI2</dt>
  <dd>Posterior weighted chi^2 P-value for testing the association
  between group1 and group2 samples.</dd>
  <dt>QCHI2</dt>
  <dd>Phred scaled PCHI2</dd>
  <dt>PR</dt>
  <dd>number of permutations yielding a smaller PCHI2.</dd>
  <dt>QBD</dt>
  <dd>Quality by Depth: QUAL/#reads</dd>
  <dt>RPB</dt>
  <dd>Read Position Bias</dd>
  <dt>MDV</dt>
  <dd>Maximum number of high-quality nonRef reads in samples</dd>
  <dt>VDB</dt>
  <dd>Variant Distance Bias (v2) for filtering splice-site
  artefacts in RNA-seq data. Note: this version may be broken.</dd>
  <dt>cluster</dt>
  <dd>variant is in a region containing many closeby variant calls</dd>
</dl>
</div>

<div id="right">
</div>

</body>
</html>

